src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,702 @@
+
+    std::cout << ">>>>>>>>>>>>> PERFCOUNTERS BEGIN <<<<<<<<<<<<" <<
+      std::endl;
+    f = Formatter::create("json-pretty");
+    g_ceph_context->get_perfcounters_collection()->dump_formatted(
+      f, false);
+    f->flush(std::cout);
+    std::cout << ">>>>>>>>>>>>>  PERFCOUNTERS END  <<<<<<<<<<<<" <<
+      std::endl;
+
+    delete f;
+  }
+
+  g_ceph_context->put();
+}
+
+static void ceph_msgr_sender_on_reply(const object_t &oid)
+{
+  struct ceph_msgr_data *data;
+  struct ceph_msgr_io *io;
+
+  /*
+   * Here we abuse object and use it as a raw pointer. Since this is
+   * only for benchmarks and testing we do not care about anything
+   * but performance.  So no need to use global structure in order
+   * to search for reply, just send a pointer and get it back.
+   */
+
+  io = (typeof(io))hex_to_ptr(oid.name);
+  data = io->data;
+  ring_buffer_enqueue(&data->io_completed_q, (void *)io);
+}
+
+
+class ReplyCompletion : public Message::CompletionHook {
+  struct ceph_msgr_io *m_io;
+
+public:
+  ReplyCompletion(MOSDOpReply *rep, struct ceph_msgr_io *io) :
+    Message::CompletionHook(rep),
+    m_io(io) {
+  }
+  void finish(int err) override {
+    struct ceph_msgr_data *data = m_io->data;
+
+    ring_buffer_enqueue(&data->io_completed_q, (void *)m_io);
+  }
+};
+
+static void ceph_msgr_receiver_on_request(struct ceph_msgr_data *data,
+					  MOSDOp *req)
+{
+  MOSDOpReply *rep;
+
+  rep = new MOSDOpReply(req, 0, 0, 0, false);
+  rep->set_connection(req->get_connection());
+
+  pthread_spin_lock(&data->spin);
+  if (data->io_inflight_nr) {
+    struct ceph_msgr_io *io;
+
+    data->io_inflight_nr--;
+    io = flist_first_entry(&data->io_inflight_list,
+			   typeof(*io), list);
+    flist_del(&io->list);
+    pthread_spin_unlock(&data->spin);
+
+    rep->set_completion_hook(new ReplyCompletion(rep, io));
+    rep->get_connection()->send_message(rep);
+  } else {
+    struct ceph_msgr_reply_io *rep_io;
+
+    rep_io = (typeof(rep_io))malloc(sizeof(*rep_io));
+    rep_io->rep = rep;
+
+    data->io_pending_nr++;
+    flist_add_tail(&rep_io->list, &data->io_pending_list);
+    pthread_spin_unlock(&data->spin);
+  }
+}
+
+class FioDispatcher : public Dispatcher {
+  struct ceph_msgr_data *m_data;
+
+public:
+  FioDispatcher(struct ceph_msgr_data *data):
+    Dispatcher(g_ceph_context),
+    m_data(data) {
+    require_authorizer = false;
+  }
+  bool ms_can_fast_dispatch_any() const override {
+    return true;
+  }
+  bool ms_can_fast_dispatch(const Message *m) const override {
+    switch (m->get_type()) {
+    case CEPH_MSG_OSD_OP:
+      return m_data->o->is_receiver;
+    case CEPH_MSG_OSD_OPREPLY:
+      return !m_data->o->is_receiver;
+    default:
+      return false;
+    }
+  }
+  void ms_handle_fast_connect(Connection *con) override {
+  }
+  void ms_handle_fast_accept(Connection *con) override {
+  }
+  bool ms_dispatch(Message *m) override {
+    return true;
+  }
+  void ms_fast_dispatch(Message *m) override {
```

Comment:
You are processing messages using always the "fast dispatch" path. This means that this function is being called directly by the IO thread of the messenger, and also this is not the most common path used by the Ceph components.

Could we have another configuration parameter for this engine to switch between using fast dispatch, or "normal" dispatch (using `ms_dispatch`), or even better have an option to specify the ratio for fast dispatch of messages?

Reply:
The fast_dispatch path should be used in cases where the message dispatch code is very quick to execute without blocking. I'm not entirely familiar with how Ceph components use these two paths, but sometimes they use the fast dispatch, and other times use the normal dispatch depending on the type of the messages received.

When processing the message in the IO thread you might be stalling the IO path, which might decrease the throughput. But because there is so much concurrency complexity involved in the messenger is hard to tell exactly the differences in behavior between dispatching directly in the IO thread or dispatching in an additional thread.

If we can control the ratio, at least we can experiment with several configurations to understand the impact on throughput and latency.